torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -2034,45 +2034,43 @@ void InitXlaModuleBindings(py::module m) {
   m.def("_get_local_shards",
         [](const std::vector<at::Tensor>& input)
             -> std::vector<std::vector<std::pair<at::Tensor, std::string>>> {
-          std::vector<runtime::ComputationClient::DataPtr> handles;
-          std::vector<at::ScalarType> element_types;
-          // Find all shard handles for transfer
+          std::vector<std::vector<std::pair<at::Tensor, std::string>>> result;
           for (auto& tensor : input) {
+            // Find all shard handles for transfer
             XLATensorPtr xtensor = bridge::GetXlaTensor(tensor);
             XLA_CHECK(xtensor->GetXlaData() != nullptr)
                 << "Shard data is not available";
-            XLA_CHECK(xtensor->sharding_spec() != nullptr)
-                << "Tensor is not sharded";
+            auto sharding_spec = xtensor->sharding_spec();
+            XLA_CHECK(sharding_spec != nullptr) << "Tensor is not sharded";
             auto handle =
                 std::dynamic_pointer_cast<runtime::ComputationClient::Data>(
                     xtensor->GetXlaData());
             std::vector<runtime::ComputationClient::DataPtr> shard_handles =
                 runtime::GetComputationClient()->GetDataShards(handle);
-            handles.insert(handles.end(), shard_handles.begin(),
-                           shard_handles.end());
+            std::vector<at::ScalarType> element_types;
             element_types.insert(element_types.end(), shard_handles.size(),
                                  MaybeUpcastToHostTorchType(
                                      shard_handles[0]->shape().element_type()));
-          }
-
-          std::vector<at::Tensor> cpu_shards =
-              XlaDataToTensors(WrapXlaData(handles), element_types);
-          // Populate the resulting vector of shards and device strings
-          std::vector<std::vector<std::pair<at::Tensor, std::string>>> result;
-          int shards_per_tensor =
-              runtime::GetComputationClient()->GetLocalDevices().size();
-          result.reserve(cpu_shards.size() / shards_per_tensor);
-          for (int i = 0; i < cpu_shards.size(); i += shards_per_tensor) {
+            std::vector<at::Tensor> cpu_shards =
+                XlaDataToTensors(WrapXlaData(shard_handles), element_types);
```

Comment:
Calling `XlaDataToTensors` on each tensor individually will slow down d2h transfers for async checkpointing, since PjRt won't be able to fully utilize transfer parallelization.

Do we expect manually-sharded tensors to contain actual device data generally, or will they usually be IR? If just IR, maybe we can add an assertion to prevent access here.

Reply:
I rather keep it functional for both cases -- shouldn't it be asynchronous anyway, not blocking the actual training run?

Reply:
This is interesting. I was not aware this performance optimization...